adapters/repos/db/vector/cache/paginated_cache.go

```diff
@@ -0,0 +1,505 @@
+	logger logrus.FieldLogger, normalizeOnRead bool, deletionInterval time.Duration,
+) Cache[float32] {
+	vc := &paginatedCache[float32]{
+		vectorForID: func(ctx context.Context, id uint64) ([]float32, error) {
+			vec, err := vecForID(ctx, id)
+			if err != nil {
+				return nil, err
+			}
+			if normalizeOnRead {
+				vec = distancer.Normalize(vec)
+			}
+			return vec, nil
+		},
+		cache:            make([][][]float32, 1),
+		count:            0,
+		maxSize:          int64(maxSize),
+		cancel:           make(chan bool),
+		logger:           logger,
+		shardedLocks:     common.NewShardedLocks(PageSize),
+		deletionInterval: deletionInterval,
+		pruneInterval:    DefaultPruneInterval,
+	}
+
+	vc.cache[0] = make([][]float32, PageSize)
+
+	vc.watchForDeletion()
+	vc.watchForPrune()
+	return vc
+}
+
+func NewPaginatedByteCache(vecForID common.VectorForID[byte], maxSize int,
+	logger logrus.FieldLogger, deletionInterval time.Duration,
+) Cache[byte] {
+	vc := &paginatedCache[byte]{
+		vectorForID:      vecForID,
+		cache:            make([][][]byte, 1),
+		count:            0,
+		maxSize:          int64(maxSize),
+		cancel:           make(chan bool),
+		logger:           logger,
+		shardedLocks:     common.NewShardedLocks(PageSize),
+		deletionInterval: deletionInterval,
+		pruneInterval:    DefaultPruneInterval,
+	}
+
+	vc.cache[0] = make([][]byte, PageSize)
+
+	vc.watchForDeletion()
+	vc.watchForPrune()
+	return vc
+}
+
+func NewPaginatedUInt64Cache(vecForID common.VectorForID[uint64], maxSize int,
+	logger logrus.FieldLogger, deletionInterval time.Duration,
+) Cache[uint64] {
+	vc := &paginatedCache[uint64]{
+		vectorForID:      vecForID,
+		cache:            make([][][]uint64, 1),
+		count:            0,
+		maxSize:          int64(maxSize),
+		cancel:           make(chan bool),
+		logger:           logger,
+		shardedLocks:     common.NewShardedLocks(PageSize),
+		deletionInterval: deletionInterval,
+		pruneInterval:    DefaultPruneInterval,
+	}
+
+	vc.cache = make([][][]uint64, 1)
+	vc.cache[0] = make([][]uint64, PageSize)
+
+	vc.watchForDeletion()
+	vc.watchForPrune()
+	return vc
+}
+
+func (s *paginatedCache[T]) All() [][]T {
+	s.shardedLocks.RLockAll()
+	defer s.shardedLocks.RUnlockAll()
+
+	out := make([][]T, 0, s.Len())
+	for _, page := range s.cache {
+		for _, vec := range page {
+			if vec != nil {
+				out = append(out, vec)
+			}
+		}
+	}
+
+	return out
+}
+
+// getPageForID returns the page and index of the vector for the given id.
+// This function assumes that the caller holds a read lock on the page.
+func (s *paginatedCache[T]) getPageForID(id uint64) (page [][]T, idx int) {
+	pageIdx := id / PageSize
+	page = s.cache[pageIdx]
+	if page == nil {
+		return nil, -1
+	}
+
+	return page, int(id % PageSize)
+}
+
+// upsert gets or creates a page and applies an update to it.
+// No locks must be held when calling this function.
+func (s *paginatedCache[T]) upsert(id uint64, update func(page [][]T, idx int) error) error {
```

Comment:
Is there benefit of having `update` callback? Only implementation I see is assigning vector to page.